reactos/drivers/usb/usbport/usbport.c

```diff
@@ -0,0 +1,2751 @@
+    {
+        DPRINT1("USBPORT_RequestAsyncCallback: Not allocated AsyncCallbackData!\n");
+        return 0;
+    }
+
+    RtlZeroMemory(AsyncCallbackData,
+                  sizeof(USBPORT_ASYNC_CALLBACK_DATA) + Length);
+
+    if (Length)
+    {
+        RtlCopyMemory(&AsyncCallbackData->CallbackContext, Buffer, Length);
+    }
+
+    AsyncCallbackData->FdoDevice = FdoDevice;
+    AsyncCallbackData->CallbackFunction = Callback;
+
+    KeInitializeTimer(&AsyncCallbackData->AsyncTimer);
+
+    KeInitializeDpc(&AsyncCallbackData->AsyncTimerDpc,
+                    USBPORT_AsyncTimerDpc,
+                    AsyncCallbackData);
+
+    DueTime.QuadPart -= (KeQueryTimeIncrement() - 1) + 10000 * TimerValue;
+
+    KeSetTimer(&AsyncCallbackData->AsyncTimer,
+               DueTime,
+               &AsyncCallbackData->AsyncTimerDpc);
+
+    return 0;
+}
+
+PVOID
+NTAPI
+USBPORT_GetMappedVirtualAddress(IN PVOID PhysicalAddress,
+                                IN PVOID MiniPortExtension,
+                                IN PVOID MiniPortEndpoint)
+{
+    PUSBPORT_COMMON_BUFFER_HEADER HeaderBuffer;
+    PUSBPORT_ENDPOINT Endpoint;
+    ULONG Offset;
+    ULONG_PTR VirtualAddress;
+
+    DPRINT_CORE("USBPORT_GetMappedVirtualAddress ... \n");
+
+    Endpoint = (PUSBPORT_ENDPOINT)((ULONG_PTR)MiniPortEndpoint -
+                                   sizeof(USBPORT_ENDPOINT));
+
+    if (!Endpoint)
+    {
+        ASSERT(FALSE);
+    }
+
+    HeaderBuffer = Endpoint->HeaderBuffer;
+
+    Offset = (ULONG_PTR)PhysicalAddress - HeaderBuffer->PhysicalAddress;
+    VirtualAddress = HeaderBuffer->VirtualAddress + Offset;
+
+    return (PVOID)VirtualAddress;
+}
+
+ULONG
+NTAPI
+USBPORT_InvalidateEndpoint(IN PVOID Context1,
+                           IN PVOID Context2)
+{
+    PUSBPORT_DEVICE_EXTENSION FdoExtension;
+    PDEVICE_OBJECT FdoDevice;
+    PUSBPORT_ENDPOINT Endpoint;
+
+    DPRINT_CORE("USBPORT_InvalidateEndpoint: ... \n");
+
+    FdoExtension = (PUSBPORT_DEVICE_EXTENSION)((ULONG_PTR)Context1 -
+                                               sizeof(USBPORT_DEVICE_EXTENSION));
+
+    FdoDevice = FdoExtension->CommonExtension.SelfDevice;
+
+    Endpoint = (PUSBPORT_ENDPOINT)((ULONG_PTR)Context2 -
+                                   sizeof(USBPORT_ENDPOINT));
+
+    if (Context2)
+    {
+        USBPORT_InvalidateEndpointHandler(FdoDevice,
+                                          Endpoint,
+                                          INVALIDATE_ENDPOINT_ONLY);
+    }
+    else
+    {
+        USBPORT_InvalidateEndpointHandler(FdoDevice,
+                                          NULL,
+                                          INVALIDATE_ENDPOINT_ONLY);
+    }
+
+    return 0;
+}
+
+VOID
+NTAPI
+USBPORT_CompleteTransfer(IN PURB Urb,
+                         IN USBD_STATUS TransferStatus)
+{
+    struct _URB_CONTROL_TRANSFER *UrbTransfer;
+    PUSBPORT_TRANSFER Transfer;
+    NTSTATUS Status;
+    PIRP Irp;
+    KIRQL OldIrql;
+    PRKEVENT Eve
```

Comment:
This whole spiel (lines 2253-2258 + line 2262) should be MmGetSystemAddressForMdlSafe